lib/internal/Magento/Framework/App/MaintenanceMode.php

```diff
@@ -127,4 +139,62 @@ public function getAddressInfo()
             return [];
         }
     }
+
+    /**
+     * Normalise / canonicalise list of IP addresses / ranges
+     *
+     * Takes a string of IP addresses or ranges (comma separated) and returns a
+     * string of IP ranges with no overlaps nor duplicates (comma separated).
+     *
+     * @param  string $addresses
+     * @return string
+     */
+    private function normaliseIPAddresses(string $addresses): string
+    {
+        $addresses = explode(',', $addresses);
+        $addresses = array_filter($addresses); // remove empty strings
+        $networks = [];
+        foreach ($addresses as $address) {
+            $networks[] = Network::parse(trim($address));
+        }
+        $networks = array_unique($networks); // remove obvious duplicates
+
+        // Combine adjacent ranges where feasible
+        if (count($networks) > 1) {
+            // Sort the list so we can compare each item with the following to
+            // determine if they are adjacent.
+            sort($networks, SORT_NATURAL);
+
+            // Define end point here as we expect the array to grow within the loop
+            $penultimate = count($networks) - 1;
+            for ($i = 0; $i < $penultimate; $i++) {
+                $a = $networks[$i];
+                $b = $networks[$i + 1];
+                if ($a->getLastIP()->next() == $b->getFirstIP()) {
+                    $range = new Range($a->getFirstIP(), $b->getLastIP());
+                    foreach ($range->getNetworks() as $network) {
+                        $networks[] = $network;
+                    }
+                }
+            }
+        }
+
+        // Remove overlapping entries from the list
+        if (count($networks) > 1) {
+            // Sort the list so we can compare each item with the following to
+            // determine if the former includes the latter.
+            sort($networks, SORT_NATURAL);
+
+            $lastRange = null;
+            $networks = array_filter($networks, function ($network) use (&$lastRange) {
+                if ($lastRange && $lastRange->contains($network)) {
+                    return false;
+                }
+                $lastRange = Range::parse($network);
+                return true;
+            });
+        }
+
+        return implode(',', $networks);
+    }
```

Comment:
This is quite complex method. Could you extract it to separate class and cover this logic with unit tests?